src/pick.c

```diff
@@ -409,41 +415,43 @@ selected_choice(void)
 		case CTRL('B'):
 		case LEFT:
 			if (cursor_position > 0)
-				--cursor_position;
+				cursor_position += utf8next(
+				    query, cursor_position, -1);
 			break;
 		case CTRL('F'):
 		case RIGHT:
 			if (cursor_position < query_length)
-				++cursor_position;
+				cursor_position += utf8next(
+				    query, cursor_position, 1);
 			break;
 		default:
-			if (key > 31 && key < 127) { /* Printable chars */
-				if (cursor_position < query_length) {
-					memmove(
-					    query + cursor_position + 1,
-					    query + cursor_position,
-					    query_length - cursor_position);
-				}
-
-				query[cursor_position++] = key;
-				query[++query_length] = '\0';
-				filter_choices();
-				selection = 0;
+			if (query_size < query_length + length) {
+				do
+					query_size *= 2;
+				while (query_size < query_length + length);
+				query = reallocarray(
+				    query, query_size, sizeof(char));
+				if (query == NULL)
+					err(1, "reallocarray");
```

Comment:
Passing `NULL` as the second argument to err(3) will print a default message, which (should, hopefully) include the text "reallocarray".